fili-core/src/main/java/com/yahoo/bard/webservice/data/filterbuilders/ConjunctionDruidFilterBuilder.java

```diff
@@ -88,6 +96,80 @@ public Filter buildFilters(Map<Dimension, Set<ApiFilter>> filterMap) throws Dime
     protected abstract Filter buildDimensionFilter(Dimension dimension, Set<ApiFilter> filters)
             throws DimensionRowNotFoundException;
 
+    /**
+     * Splits a set of {@link ApiFilter}s into two groups of positive and negative filters.
+     * <p>
+     * An {@link ApiFilter} is defined to be positive if its filter operation is one of
+     * <ul>
+     *     <li> {@link FilterOperation#in}
+     *     <li> {@link FilterOperation#startswith}
+     *     <li> {@link FilterOperation#contains}
+     *     <li> {@link FilterOperation#eq}
+     * </ul>
+     * An {@link ApiFilter} is defined to be negative if its filter operation is {@link FilterOperation#notin}.
+     *
+     * @param filters  A set of API filters that are to be splitted
+     *
+     * @return a pair whose left is positive filter and whose right is negative filers
+     */
+    protected static Pair<Set<ApiFilter>, Set<ApiFilter>> splitApiFilters(Set<ApiFilter> filters) {
+        Set<ApiFilter> positiveFilters = new HashSet<>();
+        Set<ApiFilter> negativeFilters = new HashSet<>();
+
+        for (ApiFilter filter : filters) {
+            if (FilterOperation.notin.equals(filter.getOperation())) {
+                // this is a negative filter
+                negativeFilters.add(filter);
+            } else {
+                // this is a positive filter
+                positiveFilters.add(filter);
+            }
+        }
+
+        return new ImmutablePair<>(positiveFilters, negativeFilters);
+    }
+
+    /**
+     * Negates a collection of negative filters, i.e. {@link FilterOperation#notin} {@code =>}
+     * {@link FilterOperation#in}, and returns a stream of the negated filters.
+     * <p>
+     * This method throws {@link IllegalArgumentException} if any one of the filter collection passed in is not a
+     * negative filter, i.e. {@link FilterOperation#notin}.
```

Comment:
Would it simplify this method's contract if it negated any filters passed in? So, if it gets a negative filter, it turns it into the positive filter, and if it _does_ get a positive filter, it flips it to a negative filter?

It may make the body of the method a little bit more complex, but it seems like it would greatly simplify the checking that the caller has to do to call this safely.

Alternatively, can we lean on the type system to require the collection being passed in are all negative filters?

Reply:
I was mostly hoping to move from a runtime failure detection to compile-time, but I see your point about this encapsulating the complexity. However, because it throws an illegal argument, it's actually pushing the complexity of "am I calling this with a valid set of filters" to the caller, so the caller is still forced to deal with the complexity of "make sure you're sending in the right thing". 

Having the type system enforce that makes it a little more fool-proof for the caller (even if local) because the compiler will blow up if they try to do it wrong, rather than having to build out tests to check it. Which is more complex / harder to maintain, having to have tests for all of the "scenarios" for which this method needs to look for, or only having what the type system allows?

If we've got the rich tests around the complexity of these rules, and tests ensuring the caller is handling everything correctly, I'm less concerned, but otherwise I still lean towards "harder to screw up".

Reply:
I've added additional test - https://github.com/yahoo/fili/pull/688/files#diff-dee9ab9980853004178878638e51ebe8R108